### DISCUSSION TITLE
Exclude certain treesitter captures

### DISCUSSION THREAD

**Discussion by gennaro-tedesco** [folke/flash.nvim#95] [Ideas]
I find the treesitter search and remote operator awesome, I am using it all the time. Do you think it would make sense (and if so, how difficult it is to achieve) to exclude certain treesitter captures? Take the example below

![Screenshot 2023-06-29 at 13 37 20](https://github.com/folke/flash.nvim/assets/15387611/07dd7e70-195e-4db7-af17-ad2b21274c49)

you can see that treesitter correctly identifies the word "jump" (as `identifier` capture, probably, "jump" being the word that started the search altogether): however, pragmatically, if I wanted to action on a one-word capture I would probably use standard vim objects (like `yiw` or so); as such, when I use flash treesitter search is almost always because I want to identify particular blocks of code rather than function arguments or one-line expressions. I was thinking one could introduce a configurable option for treesitter search along the lines of
```lua
treesitter_search = {
      ...
      exclude = {"identifier", "expression", "argument_list"}
      ...
    },
```
and only offer labels for blocks or objects the user is after.

Do you reckon such feature to be in scope of this plugin?

**Comment by JoseConseco** [comment]
I was just thinking the same.    The  'smallest' TSitter node is often the same as word under cursor, thus first label is redundant with . ciw, diw  - word motions.  Removing first node match would make labels cleaner I think.

**Comment by JoseConseco** [comment]
@folke  I tried to override the matcher but I'm not too good with lua.  What I got so far:

```lua
      {
        "u", -- unit textobject
        mode = { "o", "x" },
        function()
          local flash_ts = require "flash.plugins.treesitter"
          vim.print(flash_ts)
          local my_matcher = function(win, state)
            local base_results = flash_ts.matcher(win, state)
            table.remove(base_results, 1) -- remove the first result
            return base_results
          end
          flash_ts.jump {
            matcher = my_matcher,
            label = {
              rainbow = {
                enabled = true,
              },
            },
          }
        end,
      },

```
I can easily remove first node from  flash.plugins.treesitter.lua - 
  -- table.remove(ranges, 1)
And it works there, but it then breaks the  'treesitter_search' - where node directly under cursor is usefull.

**Comment by gennaro-tedesco** [comment]
I have created a rudimentary PR [here](https://github.com/folke/flash.nvim/pull/118) to show the concept.

**Comment by JoseConseco** [comment]
Here is code to remove first match (same as word under cursor) from treesitter jump (using new filter option) : 

```lua
      {
        "u", -- unit textobject
        mode = { "o", "x" },
        function()
          require("flash").treesitter {
            filter = function(matches)
              ---@param m Flash.Match.TS
              -- for all matches,  match[n+1] should use previous match[n] label. Do in reverse order, to prevet copying first label to all matches
              for i = #matches, 2, -1 do
                matches[i].label = matches[i - 1].label
              end
              table.remove(matches, 1) -- remove first match, as it is same as word under cursor thus redundant with word motion
            end,
            label = {
              rainbow = {
                enabled = true,
              },
            },
          }
        end,
      },

```
